tile/lang/ast/jacobian.cc

```diff
@@ -0,0 +1,290 @@
+// Copyright 2019 Intel Corporation.
+
+#include "tile/lang/ast/jacobian.h"
+
+#include <map>
+#include <memory>
+#include <stack>
+#include <unordered_set>
+
+#include <boost/format.hpp>
+
+#include "tile/lang/ast/gradient.h"
+#include "tile/lang/ast/traversal.h"
+
+namespace vertexai {
+namespace tile {
+namespace lang {
+namespace ast {
+
+namespace {
+
+struct UseInfo {
+  ExprPtr expr;
+  size_t idx;
+};
+
+class ComputeUses : public AstVisitor<void> {
+ public:
+  explicit ComputeUses(const ExprPtr& src) {
+    stack_.push(src);
+    while (stack_.size()) {
+      auto expr = stack_.top();
+      stack_.pop();
+      if (!seen_.count(expr.get())) {
+        seen_.insert(expr.get());
+        expr->Accept(this);
+      }
+    }
+  }
+
+  const std::vector<UseInfo>& uses(const Expr* expr) const { return uses_.at(expr); }
+
+ private:
+  void Visit(const CallExpr& expr) final {
+    for (size_t i = 0; i < expr.args.size(); i++) {
+      Push(expr, expr.args[i], i);
+    }
+  }
+
+  void Visit(const ContractionExpr& expr) final {
+    for (size_t i = 0; i < expr.srcs.size(); i++) {
+      Push(expr, expr.srcs[i]->ref, i);
+    }
+    if (expr.use_default) {
+      Push(expr, expr.use_default, expr.srcs.size());
+    }
+  }
+
+  void Visit(const GradOverrideExpr& expr) final {
+    for (size_t i = 0; i < expr.ins.size(); i++) {
+      Push(expr, expr.ins[i], i);
+    }
+  }
+
+  void Visit(const DimExprExpr& expr) final {}
+  void Visit(const FloatConst& expr) final {}
+  void Visit(const IntConst& expr) final {}
+  void Visit(const ParamExpr& expr) final {}
+
+ private:
+  void Push(const Expr& user, const ExprPtr& used, size_t idx) {
+    IVLOG(2, "ComputeUses::Push> user: " << &user << ", used: " << used << ", idx: " << idx);
+    auto ptr = std::const_pointer_cast<Expr>(user.as_ptr());
+    uses_[used.get()].emplace_back(UseInfo{ptr, idx});
+    stack_.push(used);
+  }
+
+ private:
+  std::stack<ExprPtr> stack_;
+  std::unordered_set<const Expr*> seen_;
+  std::unordered_map<const Expr*, std::vector<UseInfo>> uses_;
+};
+
+class Jacobian {
+ public:
+  explicit Jacobian(const ExprPtr& err) : uses_(err) {
+    IVLOG(2, "Gradient::Gradient> err: " << err);
+    // LogicalShape* into = LogicalShape(DataType::FLOAT32, {3});m
```

Comment:
Can this be removed?

Reply:
Yep! Sorry, missed this and a few other outdated lines/comments in my cleanup